python_modules/libraries/dagster-components/dagster_components/core/component.py

```diff
@@ -204,12 +207,55 @@ def get_component_types_in_module(
 T = TypeVar("T")
 
 
+@dataclass
+class MultiComponentsLoadContext:
+    """Context for loading an entire component hierarchy in a code location.
+    Stores resources and a cache to ensure we don't load the same component multiple times.
+    """
+
+    resources: Mapping[str, object]
+
+    def build_defs_from_component_path(self, path: Path) -> Definitions:
+        """Build a definitions object from a folder within the components hierarchy."""
+        return self._build_defs_from_component_path_inner(path.absolute().resolve())
+
+    @cached_method
+    def _build_defs_from_component_path_inner(self, path: Path) -> Definitions:
+        from dagster_components.core.component_decl_builder import path_to_decl_node
+        from dagster_components.core.component_defs_builder import defs_from_components
+
+        decl_node = path_to_decl_node(path=path)
+        if not decl_node:
+            raise Exception(f"No component found at path {path}")
+
+        context = ComponentLoadContext(
+            module_name=".".join(path.parts[-3:]),
+            decl_node=decl_node,
+            resolution_context=ResolutionContext.default(),
+            components_load_context=self,
+        )
+        with use_component_load_context(context):
+            components = decl_node.load(context)
+            return defs_from_components(
+                resources=self.resources, context=context, components=components
+            )
+
+
 @dataclass
 class ComponentLoadContext:
     module_name: str
-    resources: Mapping[str, object]
     decl_node: Optional[ComponentDeclNode]
     resolution_context: ResolutionContext
+    components_load_context: MultiComponentsLoadContext
```

Comment:
context within a context seems overly complicated. I was imagining this just being an additional property on `ComponentLoadContext`

Reply:
The lifespan of the context is only a single component load, meaning we can't stash the cache on it. The parent context exists to maintain the cache.

Reply:
Added a couple comments to distinguish the contexts' roles. If you have an idea for a clever refactor lmk